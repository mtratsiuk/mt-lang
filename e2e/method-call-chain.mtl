(def createList
  (|list l r|
    (cond
      ((= l r) list)
      (else (createList (list/concat l) (+ l 1) r)))))

(def testSum
  (~> ["1" "2" "3"]
   ~> map Number
   ~> reduce (|c r| (+ r c)) 0))

(def testDynMethod
  (~> ["1" "2" "3"]
   ~> (~> [".m." ".a." ".p."] ~> join "" ~> replaceAll "." "") Number
   ~> reduce (|c r| (+ r c)) 0))

(def testDynReceiver
  (~> (createList [] 0 10)
   ~> filter (|x| (= (% x 2) 0))
   ~> join " | "))

(print testSum)
(print testDynMethod)
(print testDynReceiver)
