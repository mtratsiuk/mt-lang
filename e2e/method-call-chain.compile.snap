const __mtl_curry = f => (...args) => {
  if (args.length >= f.length) {
    return f(...args);
  }

  return __mtl_curry(f.bind(null, ...args));
};

const createList = __mtl_curry((list, l, r) => {
  return (() => {
    if ((l === r)) {
      return list;
    } else {
      return createList(list.concat(l), (l + 1), r);
    }
  })();
});

const testSum = ["1", "2", "3"].map(Number).reduce(__mtl_curry((c, r) => {
  return (r + c);
}), 0);

const testDynMethod = ["1", "2", "3"][[".m.", ".a.", ".p."].join("").replaceAll(".", "")](Number).reduce(__mtl_curry((c, r) => {
  return (r + c);
}), 0);

const testDynReceiver = createList([], 0, 10).filter(((x) => {
  return ((x % 2) === 0);
})).join(" | ");

console.log(String(testSum));

console.log(String(testDynMethod));

console.log(String(testDynReceiver));
